Float16 (#1120)
apple/swift-evolution: proposals/0000-float16.md

# Float16

* Proposal: [SE-0000](0000-float16.md)
* Author: [Stephen Canon](https://github.com/stephentyrone)
* Review Manager: TBD
* Pitch thread: [Add Float16](https://forums.swift.org/t/add-float16/33019)
* Implementation: [Implement Float16](https://github.com/apple/swift/pull/21738)
* Status: **Awaiting review**

## Introduction

Introduce the `Float16` type conforming to the `BinaryFloatingPoint` and `SIMDScalar`
protocols, binding the IEEE 754 *binary16* format (aka *float16*, *half-precision*, or *half*),
and bridged by the compiler to the C `_Float16` type.

* Old pitch thread: [Add `Float16`](https://forums.swift.org/t/add-float16/19370).

## Motivation

The last decade has seen a dramatic increase in the use of floating-point types smaller

> than (32-bit) `Float`. The most widely implemented is `Float16`, which is used

Reviewer: It might be clearer to use `Float32` and `Float64` throughout.

```suggestion
than `Float32`. The most widely implemented is `Float16`, which is used
```

Author: I think most people reviewing the proposal are more familiar with the names `Float` and `Double` (indeed, several people were surprised that the other names exist).

extensively on mobile GPUs for computation, as a pixel format for HDR images, and as
a compressed format for weights in ML applications.

Introducing the type to Swift is especially important for interoperability with shader-language
programs; users frequently need to set up data structures on the CPU to
pass to their GPU programs. Without the type available in Swift, they are forced to use
unsafe mechanisms to create these structures.

In addition, C APIs that use these types simply cannot be imported, making them 
unavailable in Swift.

## Proposed solution

Add `Float16` to the standard library.

## Detailed design

There is shockingly little to say here. We will add:
```
@frozen
struct Float16: BinaryFloatingPoint, SIMDScalar {
  // ...
}
```
The entire API falls out from that. I.e. `Float16` will provide exactly the operations that
`Float` and `Double` and `Float80` do for their conformance to these protocols.

The nitty-gritty of implementation is slightly more involved; we need to update a few
python files that drive `.gyb` generation, to introduce availability annotations for the type

> and operations (all the other floating-point types are unconditionally availble, so there's

Reviewer: Typo: `availble` ➡️ `available`

```suggestion
and operations (all the other floating-point types are unconditionally available, so there's
```

more people come forward, however, we should make the "opinionated" choice to have a single
name for the type. An alias could always be added with a subsequent minor proposal if
necessary.

Q: What about ARM's ["alternative half precision"](https://en.wikipedia.org/wiki/Half-precision_floating-point_format#ARM_alternative_half-precision)?
What about [bfloat16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format)?

A: Alternative half-precision is no longer supported; ARMv8.2-FP16 only uses the IEEE 754
encoding. Bfloat is something that we should *also* support eventually, but it's a separate
proposal. and we should wait a little while before doing it. Conformance to IEEE 754 fully
defines the semantics of `Float16` (and several hardware vendors, including Apple have
been shipping processors that implement those semantics for a few years). By contrast,
there are a few proposals for hardware that implements bfloat16; ARM and Intel designed
different multiply-add units for it, and haven't shipped yet (and haven't defined any 
operations *other* than a non-homogeneous multiply-add). Other companies have HW in

> use, but haven't (publically) formally specified their arithmetic. It's a moving target, and it

Reviewer: Typo: `publically` ➡️ `publicly`

```suggestion
use, but haven't (publicly) formally specified their arithmetic. It's a moving target, and it
```
